pkg/app/testdata/testapply/noop/log

```diff
@@ -34,4 +34,11 @@ second-pass rendering result of "helmfile.yaml.part.0":
 merged environment: &{default map[] map[]}
 2 release(s) found in helmfile.yaml
 
+invoking preapply hooks for 2 groups of releases in this order:
+GROUP RELEASES
+1     default//foo
+2     default//bar
+
+invoking preapply hooks for releases in group 1/2: default//foo
+invoking preapply hooks for releases in group 2/2: default//bar
```

Comment:
In this case, is this the expected value?

Reply:
Yeah probably. I thought this is an example of what we call "Release ID" in helmfile code. The first part is the kubecontext and the second part is a K8s namespace, followed by the release name. `//` looks like there were no namespace specified. Try setting `namespace: whatever` in the release so that the release ID would become `default/whatever/bar`.

Note that making it appear like `default/bar` when you omitted the release namespace doesn't look right, because then you have no way to distinguish a release whose namespace is omitted vs another release whose kubecontext is omitted.

Reply:
There is also a preapply problem. In this case, the chart has been declared not to be installed. 
**case**
```
releases:
- name: bar
  chart: stable/mychart2
- name: foo
  chart: stable/mychart1
  installed: false
  needs:
  - bar
```
**before**
`preapply` will not be executed, I think it is reasonable

```
processing file "helmfile.yaml" in directory "."
changing working directory to "/path/to"
first-pass rendering starting for "helmfile.yaml.part.0": inherited=&{default map[] map[]}, overrode=<nil>
first-pass uses: &{default map[] map[]}
first-pass rendering output of "helmfile.yaml.part.0":
 0: 
 1: releases:
 2: - name: bar
 3:   chart: stable/mychart2
 4: - name: foo
 5:   chart: stable/mychart1
 6:   installed: false
 7:   needs:
 8:   - bar
 9: 

first-pass produced: &{default map[] map[]}
first-pass rendering result of "helmfile.yaml.part.0": {default map[] map[]}
vals:
map[]
defaultVals:[]
second-pass rendering result of "helmfile.yaml.part.0":
 0: 
 1: releases:
 2: - name: bar
 3:   chart: stable/mychart2
 4: - name: foo
 5:   chart: stable/mychart1
 6:   installed: false
 7:   needs:
 8:   - bar
 9: 

merged environment: &{default map[] map[]}
2 release(s) found in helmfile.yaml

changing working directory back to "/path/to"
```

**now**
`preapply` will be executed, which is dangerous

**more reasonable**
Detect a release that needs to be installed and execute `preapply`. Whether this release needs to be updated or not.

@mumoshu @yxxhero @itscaro WDYT?

Reply:
In this PR, we opted to make `preapply` happen even for releases that have no diff, right?
Then I'd guess it would also be reasonable to trigger `preapply` also for releases with `installed: false`, as the setting is intended to mean that "it's going to be deleted", not "exclude from the targeted releases".

In addition to that- could you elaborate a bit more on how it can be "dangerous"? Depending on the use case you had in your mind, we could discuss how to address it in another way.

Reply:
```
environments:
  test01:
  test02:

releases:
- name: bar
  chart: stable/mychart2
- name: foo
  chart: stable/mychart1
  installed: {{ eq .Environment.Name "production" test01 }}
  hooks:
    - events: ["preapply"]
      showlogs: true
      command: "echo"
      args: ["Hello from preapply"]
```

In this case, `helmfile -e test02 apply` will not install `foo`, but will execute `preapply` in `foo`, which is very dangerous if `preapply` has modifications to the cluster environment

Reply:
It is a problem.

Reply:
Even the `prepare` hook is dangerous if you made it to conduct a "dangerous" operation on a cluster. The original motivation to add `preapply` was to add another extension point that is superior to existing hooks. `presync` doesn't get triggered on unmodified releases, but we needed `preapply` for use cases where you need it to be triggered on unmodified releases.

Reply:
My advice would be to design your `preapply` hook command not to break your cluster when it's called on unmodified releases. Same for `prepare`.
Note also that whatever hook command should be made "idempotent", so that it is safe for rerunning. Helmfile doesn't guarantee that e.g. `preapply` is called only once for your release. For example, if helm-upgdate command failed after a `preapply` hook, you'd need to rerun the whole helmfile command and it would result in rerunning your `preapply` hook. If you've designed your preapply any dangerous operation, it would fail for the second time. That's an issue in your preapply hook, not Helmfile's preapply design.

Reply:
@xiaomudk @mumoshu when a release will not be installed, should we auto-disable the hooks?

Reply:
I think we'd better NOT disable hooks in that case. `preapply` is supposed to get triggered even on unmodified releases. It should work the same for releases with `installed: false`. `installed: false` declares the release to be deleted, unlike when you filtered out the release by using a label selector so that it is ever installed.
Indeed, we shouldn't call hooks for releases that are filtered out by using a label selector. But I think that's already implemented.

Reply:
I very much agree that hook command should be made "idempotent", I think it would be better to state this in the docs, I'll do it.